### ISSUE TITLE
Google Pub/Sub gRPC StreamingPull speed limit

### ISSUE THREAD

**Issue by lustefaniak** [akka/alpakka#2101]
### Short description

It would be useful to allow running multiple StreamingPull requests in parallel in `subscribe` method. Each request is limited to 10 MB/s per open stream.

### Details

During migration from our home-made Akka Streams wrapper around Google libs we noticed much lower throughput of our services. Problems happens only(or at least is most visible) on topics with very large messages, eg. 300-500KB. 
After investigation we are certain it is caused by 10Mb/s limit enforced per connection by Cloud Pub/Sub api.  

https://cloud.google.com/pubsub/quotas#resource_limits

When running multiple subscriptions and merging their results we achieve 4x increase of throughput which matches 4 streams we use.

What would be best way to implement that change in Alpakka? 

I think that best approach would be to scale upstream sources according to demand but only if there are messages to be consumed, but tbh I have no idea how to do that without custom, complex GraphStage. New implementation should also handle retryable errors reported in https://github.com/akka/alpakka/issues/1615

**Comment by seglo** [comment]
Hi @lustefaniak.  That sounds like an interesting idea.

Are you suggesting a solution that runs multiple `StreamingPullRequests` in parallel based on the same request?  I'm not very familiar with Google PubSub.  Is there a way to inspect the pull request and split it into multiple pull requests that return different "partitions" of data?  Or would issuing the same subscribe request automatically group partitions across streams, sort of like Consumer Groups in Kafka?

There are a few ways to implement parallel streams.  One would be to asynchronously emit one `Source` per request downstream, and then let the user consume each stream independently, or allow them to merge them together on their own.  This is similar to how [Alpakka Kafka's partitioned](https://doc.akka.io/docs/alpakka-kafka/current/consumer.html#source-per-partition) sources work.  Another option would be to use [Akka Streams sub-streams](https://doc.akka.io/docs/akka/current/stream/stream-substream.html).  You could create the sub-stream by using the `groupBy` operator to create the number of sub-streams you want and then emit a new PubSub `Source` subscription in each.  I think it would be possible to implement either without the need for a custom graph stage.

**Comment by lustefaniak** [comment]
From what I can tell, there are no clear rules which subscription request will get what messages, all is decided by the API implementation and even docs are a little bit cryptic about that.

Subscriptions in Cloud Pub/Sub work like consumer groups in Kafka, main difference is that offset is not per partition, but each message is acknowledged separately by its ACK identifier and server decides when to re-deliver the message again. Also from what I can tell server decides to which request it will be delivered and there is no control over that.

I will have to check if for small messages having single request makes any difference compared to multiple `Source`s merged together. 

My idea of having dynamic number of sources is based on possible overhead of handling multiple sources and having multiple connections open. Unfortunately I never worked with gRPC so I have not much knowledge if different request creates or reuses connection from pool or is just an HTTP2 stream in existing connection, in that case it is hard to tell how throttling is enforced.

When you send single StreamingPullRequest and in subscription there are no messages waiting it will eventually fail the source with some retryable error as in https://github.com/akka/alpakka/issues/1615. In such case there is no point opening separate request as they will also fail around same time. Similar problem happens when you consume all messages and there are no new - this will also fail `subscribe` after wait time by RST from server.

This manual error handling with RestartSource + stream merging makes realworld usage of this connector much harder than it should be.

**Comment by seglo** [comment]
> Subscriptions in Cloud Pub/Sub work like consumer groups in Kafka, main difference is that offset is not per partition, but each message is acknowledged separately by its ACK identifier and server decides when to re-deliver the message again. Also from what I can tell server decides to which request it will be delivered and there is no control over that.

If there's something like a Kafka `group.id` then I would imagine it works the same way.  Each `StreamingPullRequest` would receive distinct records.  Maybe you can verify if that's the case.

> My idea of having dynamic number of sources is based on possible overhead of handling multiple sources and having multiple connections open. Unfortunately I never worked with gRPC so I have not much knowledge if different request creates or reuses connection from pool or is just an HTTP2 stream in existing connection, in that case it is hard to tell how throttling is enforced.

This probably largely depends on the implementation of the gRPC client.  It's not immediately clear to me if there's any kind of connection pooling logic.  I don't see any at first glance.

https://github.com/googleapis/java-pubsub/tree/master/proto-google-cloud-pubsub-v1/

If there isn't we can easily control how many streams we want to create in parallel using Akka Streams asynchronous operators, like `mapAsync`.

> This connector is extremely useful and it is great that it integrates so nicely with rest of Akka Streams.

That's great to hear.  We would welcome any contributions you could make to make it better!
